src/main/java/com/gregtechceu/gtceu/api/gui/widget/TankWidget.java

```diff
@@ -595,24 +602,36 @@ private int tryClickContainer(boolean isShiftKeyDown) {
                             player.position().z, SoundEvents.BUCKET_EMPTY, SoundSource.BLOCKS, 1.0f, 1.0f);
 
                 }
-                gui.getModularUIContainer().setCarried(currentStack);
-                return currentStack.getCount();
+                if (currentStack.isEmpty()) {
+                    gui.getModularUIContainer().setCarried(filledResult);
+                } else {
+                    gui.getModularUIContainer().setCarried(currentStack);
+                    player.getInventory().placeItemBackInInventory(filledResult);
+                }
+                return gui.getModularUIContainer().getCarried().getCount();
             }
         }
 
         if (allowClickDrained) {
             boolean performedEmptying = false;
+            ItemStack drainedResult = ItemStack.EMPTY;
             for (int i = 0; i < maxAttempts; i++) {
                 FluidActionResult result = FluidUtil.tryEmptyContainer(currentStack, fluidTank, Integer.MAX_VALUE, null,
                         false);
                 if (!result.isSuccess()) break;
                 ItemStack remainingStack = FluidUtil
                         .tryEmptyContainer(currentStack, fluidTank, Integer.MAX_VALUE, null, true).getResult();
-                currentStack.shrink(1);
                 performedEmptying = true;
-                if (!remainingStack.isEmpty() && !player.getInventory().add(remainingStack)) {
-                    Block.popResource(player.level(), player.getOnPos(), remainingStack);
-                    break;
+
+                currentStack.shrink(1);
+
+                if (drainedResult.isEmpty())
+                    drainedResult = remainingStack.copy();
+                else if (ItemStack.isSameItemSameTags(drainedResult, remainingStack))
+                    drainedResult.grow(1);
+                else {
+                    player.getInventory().placeItemBackInInventory(drainedResult);
+                    drainedResult = remainingStack.copy();
```

Comment:
Last branch of if/else chain has brackets, so all branches should have brackets for clarity